### PR TITLE
Pyclient upate draft

### PR DESCRIPTION
**Pull request by PythonFZ**

### CODE REVIEW

File: zndraw/server/tasks.py [hunk 1]
```diff
@@ -604,3 +599,39 @@ def run_modifier(url: str, token: str, data: dict):
...
     return
+
+
+@shared_task
+@typecast
+def handle_room_get(data: RoomGetData, token: str, url: str, target: str):
+    from zndraw.zndraw_worker import ZnDrawWorker
+
+    print("I AM RUNNING THE CELERY TASK")
+    print(50 * "-")
```
Comment (PythonFZ) [inline zndraw/server/tasks.py:610]: remove

File: zndraw/utils.py [hunk 2]
```diff
@@ -23,6 +24,12 @@ def split_list_into_chunks(lst, n):
         yield lst[i : i + n]
 
 
+def estimate_max_batch_size_for_socket(frames: list[znframe.Frame], max_size=1e6):
+    sizes = [sys.getsizeof(frame) for frame in frames]
+    largest_frame = max(sizes)
+    return int(max_size / largest_frame * 0.9)
```
Comment (PythonFZ) [inline zndraw/utils.py:30]: we should get the `max_size` from the config e.g. if we want to adapt the socketio settings

File: zndraw/zndraw.py [hunk 3]
```diff
@@ -53,13 +59,16 @@ class ZnDrawBase:  # collections.abc.MutableSequence
     """
 
     url: str
-    token: str = None
+    token: str | None = None
```
Comment (PythonFZ) [inline zndraw/zndraw.py:62]: these are already in the client.

File: zndraw/zndraw.py [hunk 4]
```diff
@@ -102,46 +111,43 @@ def reconnect(self) -> None:
...
         log.critical("Reconnected to server")
 
+    @typecast_kwargs
+    def get_data(self, **data: RoomGetData) -> RoomGetData:
+        with self._lock:
+            self._data = None
+            self.socket.emit("room:get", data.to_dict())
+            while self._data is None:
+                self.socket.sleep()
+                # generous timeout
```
Comment (PythonFZ) [inline zndraw/zndraw.py:121]: I assume `self.socket.sleep()` defaults to 1 second? I like the explicit arg though

File: zndraw/zndraw.py [hunk 5]
```diff
@@ -102,46 +111,43 @@ def reconnect(self) -> None:
...
-                    FrameData(
-                        index=index,
-                        data=value.to_dict(built_in_types=False),
-                        update=True,
-                        update_database=True,
-                    )
-                )
-            ],
+        self.set_data(
+            {index: value.to_dict(built_in_types=False)}, update_database=True
```
Comment (PythonFZ) [inline zndraw/zndraw.py:139]: This is missing a `frames={}` I think

File: zndraw/zndraw.py [hunk 6]
```diff
@@ -102,46 +111,43 @@ def reconnect(self) -> None:
...
-            index = [index] if isinstance(index, int) else index
-            index = [i if i >= 0 else length + i for i in index]
-
-            data = {"index": index, "token": self.token}
-
-            self.socket.emit("atoms:delete", data)
-            if not is_slice and (index[0] >= length or index[-1] >= length):
-                raise IndexError("Index out of range")
+            index = self.wrap_and_check_index(index, length)
+            self.set_data({i: None for i in index}, update_database=True)
```
Comment (PythonFZ) [inline zndraw/zndraw.py:150]: also missing the `frames={}` I guess

File: zndraw/zndraw.py [hunk 7]
```diff
@@ -233,11 +219,11 @@ def log(self, message: str) -> None:
     @property
     def atoms(self) -> ase.Atoms:
         """Return the atoms at the current step."""
-        return self[self.step]
+        return self[self.step]  # type: ignore
 
     @property
     def points(self) -> np.ndarray:
-        data = self.socket.call("points:get", timeout=self.config.call_timeout)
+        data = self.get_data(points=True).points
```
Comment (PythonFZ) [inline zndraw/zndraw.py:226]: We need to check type consistency here

File: zndraw/zndraw.py (hunk 5 above)
Comment (RokasEl) [reply zndraw/zndraw.py:139]: Good catch!

File: zndraw/utils.py (hunk 2 above)
Comment (RokasEl) [reply zndraw/utils.py:30]: Sure, I assume we also need to configure the socket somewhere in the `app.py` for this as well

**Comment by PythonFZ** [comment]
superseded by https://github.com/zincware/ZnDraw/pull/398

**Pull request closed without merging**